### PR TITLE
Fix for 'Uninitialized string offset: 0' when using Request's `referer` under PHP 7

### LINKED ISSUE #8794: 'Uninitialized string offset: 0' when using request->referer coming from home page (/) under php7
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.2.8
* Platform and Target: PHP 7 on Ubuntu Linux

### What you did

Use `$this->request->referer(true)` when the referer is our home (/) crashes under PHP7 with the error `Uninitialized string offset: 0`.

The problem is in this line of the `Request@referer` method:

~~~php
if ($local && strpos($ref, $base) === 0) {
    $ref = substr($ref, strlen($base));
    if ($ref[0] !== '/') {
        $ref = '/' . $ref;
    }
    return $ref;
}
~~~

`$ref` is an empty string due to the `substr($ref, strlen($base))`.

### Expected Behavior

`referer` should be returning `/`

### Actual Behavior

Crashes with the error  `Uninitialized string offset: 0`.

I'm just creating a PR with the fix and its test. BTW, I'm not sure about how I've fixed, but let's discuss it on the PR.

### PR DESCRIPTION
**Pull request by elboletaire**
Fixes #8794 

As said in the issue, I'm not sure about how I've fixed it. But let me know what could I change to improve it.

### CODE REVIEW

**Comment by elboletaire** [comment]
Now I like it better. Let me know what do you think :smiley:

**Comment by ADmad** [comment]
This issue looks similar to the fix done for 2.x https://github.com/cakephp/cakephp/commit/3ad68db5ebb71de3ed97950977ed6b452cca359a

That code looks much cleaner.

**Comment by elboletaire** [comment]
Yup, seems exactly the same issue. With `that code` you mean the one by @markstory, right?

If so, let me change it.

**Comment by ADmad** [comment]
Yes I meant the code in commit I linked :)

**Comment by elboletaire** [comment]
Done :)

**Comment by ADmad** [comment]
Please squash the commits.

**Comment by elboletaire** [comment]
Sure. BTW, on my first commit I was checking for empty and wasn't necessary. Something like this would be enough:

~~~php
$ref = substr($ref, strlen($base)) ?: '/';
if ($ref[0] !== '/') {
    $ref = '/' . $ref;
}
~~~ 

pd. I'm squashing

**Comment by elboletaire** [comment]
Done :)

File: src/Network/Request.php [hunk 1]
```diff
@@ -547,6 +547,9 @@ public function referer($local = false)
         if (!empty($ref) && !empty($base)) {
             if ($local && strpos($ref, $base) === 0) {
                 $ref = substr($ref, strlen($base));
+                if (empty($ref)) {
```
Comment (markstory) [inline src/Network/Request.php]: Perhaps we should use `strlen()`? `empty()` eats `'0`'.

Comment (elboletaire) [inline src/Network/Request.php]: Yup, you're right. I've changed it in a52a8d1 adding an assertion for it too.

**Comment by markstory** [comment]
Thanks :+1:

**Comment by elboletaire** [comment]
You're welcome, as always! :smiley:

**Pull request merged**